include/util/cell_storage.hpp

```diff
@@ -0,0 +1,295 @@
+#ifndef OSRM_UTIL_CELL_STORAGE_HPP
+#define OSRM_UTIL_CELL_STORAGE_HPP
+
+#include "util/assert.hpp"
+#include "util/multi_level_partition.hpp"
+#include "util/typedefs.hpp"
+#include "util/for_each_range.hpp"
+
+#include <algorithm>
+#include <numeric>
+#include <utility>
+#include <vector>
+
+namespace osrm
+{
+namespace util
+{
+
+class CellStorage
+{
+  public:
+    using WeightOffset = std::uint32_t;
+    using BoundaryOffset = std::uint32_t;
+    using BoundarySize = std::uint32_t;
+    using SourceIndex = std::uint32_t;
+    using DestinationIndex = std::uint32_t;
+
+    static constexpr auto INVALID_WEIGHT_OFFSET = std::numeric_limits<WeightOffset>::max();
+    static constexpr auto INVALID_BOUNDARY_OFFSET = std::numeric_limits<BoundaryOffset>::max();
+
+  private:
+    struct CellData
+    {
+        WeightOffset weight_offset = INVALID_WEIGHT_OFFSET;
+        BoundaryOffset source_boundary_offset = INVALID_BOUNDARY_OFFSET;
+        BoundaryOffset destination_boundary_offset = INVALID_BOUNDARY_OFFSET;
+        BoundarySize num_source_nodes = 0;
+        BoundarySize num_destination_nodes = 0;
+    };
+
+    // Implementation of the cell view. We need a template parameter here
+    // because we need to derive a read-only and read-write view from this.
+    template <typename WeightPtrT> class CellImpl
+    {
+      private:
+        using WeightValueT = EdgeWeight;
+        using WeightRefT = decltype(*WeightPtrT());
+        BoundarySize num_source_nodes;
+        BoundarySize num_destination_nodes;
+
+        WeightPtrT const weights;
+        const NodeID *const source_boundary;
+        const NodeID *const destination_boundary;
+
+        using RowIterator = WeightPtrT;
+        class ColumnIterator : public std::iterator<std::random_access_iterator_tag, EdgeWeight>
```

Comment:
On second thought, shouldn't a column iterator be the same as a strided iterator over your flattened matrix?

http://www.boost.org/doc/libs/1_55_0/libs/range/doc/html/range/reference/adaptors/reference/strided.html